// hexo-inject/lib/util.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

exports.camalize = camalize;
exports.callsite = callsite;

var _const = require('./const');

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function camalize(str) {
  return str.split('_').filter(function (s) {
    return s.length > 0;
  }).map(function (s, i) {
    return i === 0 ? s : s[0].toUpperCase() + s.substr(1);
  }).join('');
}

function callsite() {
  function parse(t) {
    var _REGEX$stack_trace$ex = _const.REGEX.stack_trace.exec(t);

    var _REGEX$stack_trace$ex2 = _slicedToArray(_REGEX$stack_trace$ex, 6);

    var functionName = _REGEX$stack_trace$ex2[1];
    var alias = _REGEX$stack_trace$ex2[2];
    var filePath = _REGEX$stack_trace$ex2[3];
    var line = _REGEX$stack_trace$ex2[4];
    var col = _REGEX$stack_trace$ex2[5];

    var file = _path2.default.parse(filePath);

    line = parseInt(line);
    col = parseInt(col);

    return {
      functionName: functionName, alias: alias,
      filePath: filePath, file: file,
      line: line, col: col
    };
  }
  var stack = new Error().stack.split('\n').slice(2) // First line is 'Error', second line is this function
  .map(parse);

  return stack;
}
