// hexo-inject/lib/mixin/content.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

require('babel-polyfill');

var _underscore = require('underscore');

var _underscore2 = _interopRequireDefault(_underscore);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function resolve(o) {
  for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    args[_key - 1] = arguments[_key];
  }

  if (typeof o === 'function') return o.apply(undefined, args);
  return o;
}

var Content = {
  _resolveContent: function _resolveContent(src, _ref) {
    var _this = this;

    var html = _ref.html;
    var opts = _ref.opts;
    return (0, _bluebird.coroutine)(regeneratorRuntime.mark(function _callee() {
      var shouldInject;
      return regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              opts = opts || { shouldInject: true };
              html = resolve(html, src);
              shouldInject = resolve(opts.shouldInject, src);
              _context.next = 5;
              return _bluebird2.default.props({ html: html, shouldInject: shouldInject });

            case 5:
              return _context.abrupt('return', _context.sent);

            case 6:
            case 'end':
              return _context.stop();
          }
        }
      }, _callee, _this);
    }))();
  },
  _resolveInjectionPoint: function _resolveInjectionPoint(src, pos) {
    var _this2 = this;

    return (0, _bluebird.coroutine)(regeneratorRuntime.mark(function _callee2() {
      return regeneratorRuntime.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              _context2.next = 2;
              return _bluebird2.default.map(_this2._injectors[pos], _this2._resolveContent.bind(_this2, src));

            case 2:
              return _context2.abrupt('return', _context2.sent);

            case 3:
            case 'end':
              return _context2.stop();
          }
        }
      }, _callee2, _this2);
    }))();
  },
  _buildHTMLTag: function _buildHTMLTag(name, attrs, content, endTag, src) {
    var _this3 = this;

    return (0, _bluebird.coroutine)(regeneratorRuntime.mark(function _callee3() {
      var _ref2, _ref3, attr_list, html;

      return regeneratorRuntime.wrap(function _callee3$(_context3) {
        while (1) {
          switch (_context3.prev = _context3.next) {
            case 0:
              _context3.next = 2;
              return _bluebird2.default.all([_bluebird2.default.props(_underscore2.default.mapObject(attrs, function (value) {
                return resolve(value, src);
              })), resolve(content || '', src)]);

            case 2:
              _ref2 = _context3.sent;
              _ref3 = _slicedToArray(_ref2, 2);
              attrs = _ref3[0];
              content = _ref3[1];
              attr_list = _underscore2.default.map(attrs, function (value, key) {
                return key + '=\'' + value + '\'';
              }).join(' ');
              html = '<' + name + ' ' + attr_list + '>' + (endTag ? content + '</' + name + '>' : '');
              return _context3.abrupt('return', html);

            case 9:
            case 'end':
              return _context3.stop();
          }
        }
      }, _callee3, _this3);
    }))();
  },
  raw: function raw(pos, html, opts) {
    this._injectors[pos].push({ html: html, opts: opts });
  },
  tag: function tag(pos, name, attrs, content, endTag, opts) {
    return this.raw(pos, this._buildHTMLTag.bind(this, name, attrs, content, endTag), opts);
  },
  script: function script(pos, attrs, content, opts) {
    return this.tag(pos, 'script', attrs, content, true, opts);
  },
  style: function style(pos, attrs, content, opts) {
    return this.tag(pos, 'style', attrs, content, true, opts);
  },
  link: function link(pos, attrs, opts) {
    return this.tag(pos, 'link', attrs, "", false, opts);
  }
};

exports.default = Content;
module.exports = exports['default'];
